### ISSUE TITLE
Test app needs to save screen location across recreation

### ISSUE THREAD

**Issue by JakeWharton** [cashapp/redwood#1518]
So a config change puts you back onto the same screen rather than the home screen.

**Comment by ahmed3elshaer** [comment]
For some reason using `rememberSaveable` doesn't work properly.
Inside the `TestApp`
I replaced 
```kotlin
  val screenKeyState = remember { mutableStateOf<String?>(null) }
```
with 
```kotlin
  val screenKeyState = rememberSaveable { mutableStateOf<String?>(null) }
```
And still when navigating to `Repo Search` screen and change rotation i get back to `HomeScreen`

For the sake of debugging i provided the `Saver` and added some print lines
```kotlin
  val screenKeyState = rememberSaveable(
    saver = Saver(
      save = {
        println("${it.value} screen is saved")
        it
      },
      restore = {
        println("${it.value} screen is restored")
        it
      },
    ),
  ) { mutableStateOf<String?>(null) }

  val screenKey = screenKeyState.value
  println("screenKey: $screenKey")
```
And i noticed while navigating to `Repo Search` the `save` lambda doesn't get called.
here is the Logcat
```
screenKey: null
null screen is saved
screenKey: Repo Search
Rotation=========
screenKey: null
null screen is saved
```
Any thoughts ? 🤔

**Comment by JakeWharton** [comment]
Honestly I don't know anything about how it works. Maybe #1531 or #1533 is the fix?

**Comment by ahmed3elshaer** [comment]
Maybe, Specially #1533 
I will test again when it's merged and let you know

**Comment by ahmed3elshaer** [comment]
I have tested the app and found that it does not preserve the state across configurations.
 I have debugged the issue and found that `onSaveInstanceState` is not invoked in `TreehouseLayout` when navigating to a new screen. Moreover, the app crashed once when rotating the device and here are the logs:
```
09:17:26.097 Parcel             Reading a NULL string not supported here.
09:17:26.098 Parcel             Reading a NULL string not supported here.
09:17:26.099 Parcel             Reading a NULL string not supported here.
09:17:26.100 Parcel             Reading a NULL string not supported here.
09:17:26.100 Parcel             Reading a NULL string not supported here.
09:17:26.101 Parcel             Reading a NULL string not supported here.
09:17:26.102 Parcel             Reading a NULL string not supported here.
09:17:27.666 AndroidRuntime     FATAL EXCEPTION: Treehouse
                                Process: com.example.redwood.testing.android.views, PID: 21992
                                app.cash.zipline.ZiplineException: IllegalStateException: unexpected type: null
                                	at captureStack (runtime/coreRuntime.kt:12)
                                	at IllegalStateException_init_$Create$_0 (kotlin-kotlin-stdlib-js-ir.js)
                                	at toJsonElement (redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt:96)
                                	at toStateSnapshot (commonMainSources/libraries/stdlib/src/kotlin/collections/Maps.kt:58)
                                	at <anonymous> (redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt:114)
                                	at <anonymous> (redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/ZiplineTreehouseUi.kt)
                                	at <anonymous> (redwood-redwood-treehouse.js)
                                	at <anonymous> (runner/work/zipline/zipline/zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/InboundService.kt:1389)
                                	at <anonymous> (runner/work/zipline/zipline/zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt:99)
                                	at <anonymous> (zipline-root-zipline.js)
                                	at <anonymous> (runner/work/zipline/zipline/zipline/src/jsMain/kotlin/app/cash/zipline/GlobalBridge.kt)
                                	at <anonymous> (zipline-root-zipline.js)
                                	at app.cash.redwood.treehouse.ZiplineTreehouseUi$Companion$Adapter$GeneratedOutboundService.snapshotState(ZiplineTreehouseUi.kt:29)
                                	at app.cash.redwood.treehouse.ViewContentCodeBinding$performSave$1.invokeSuspend(TreehouseAppContent.kt:440)
                                	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
                                	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
                                	at java.lang.Thread.run(Thread.java:1012)
                                	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [app.cash.redwood.compose.AndroidUiFrameClock@e8443fb, StandaloneCoroutine{Cancelling}@a299c18, java.util.concurrent.Executors$FinalizableDelegatedExecutorService@3d6f671]
```

**Comment by veyndan** [comment]
So I think the issue is that the default `StateStore` implementation is an in-memory state store ([`MemoryStateStore`](https://github.com/cashapp/redwood/blob/trunk/redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/MemoryStateStore.kt)). This gets instantiated as part of the `TreehouseAppFactory`, which in the `TestApp` is instantiated [within `onCreate` of the `TestAppActivity`](https://github.com/cashapp/redwood/blob/5c2c4d1b474c0ec097b5f9204aa3cab9e38432d3/test-app/android-views/src/main/kotlin/com/example/redwood/testing/android/views/TestAppActivity.kt#L80). All this is to say that the state store is reinstantiated whenever there is a config change, which for an in-memory state store means that the data is wiped whenever there is a config change.

The Emoji Search sample app gets around this by creating a custom `StateStore` instance that writes to a file (i.e., [`FileStateStore`](https://github.com/cashapp/redwood/blob/trunk/samples/emoji-search/android-views/src/main/kotlin/com/example/redwood/emojisearch/android/views/FileStateStore.kt)). Maybe we want something similar? Though I feel like we'd probably want a better default `StateStore` implementation than one that doesn't work out of the box…

**Comment by ahmed3elshaer** [comment]
@veyndan thank you for your valuable insight. It was helpful to understand the reason behind the issue. I believe I can implement a solution similar to `FileStateStore` to ensure safety.

Do you think it would be beneficial to open another issue for a default `StateStore` implementation that survives config change by default?

Furthermore, I’m curious about the tradeoffs if we were to default `FileStateStore` over an in-memory store? 🤔

**Comment by veyndan** [comment]
> Do you think it would be beneficial to open another issue for a default `StateStore` implementation that survives config change by default?
> 
> Furthermore, I’m curious about the tradeoffs if we were to default `FileStateStore` over an in-memory store? 🤔

I think that'd be awesome! I don't know if defaulting to a `FileStateStore` is the ideal solution though — it's just an existing `StateStore` that supports values surviving config change. Perhaps we need a completely different implementation of `StateStore` to be the default? 🤷 

I think a great first step is to create a separate issue so we have a place to put ideas!

**Comment by ahmed3elshaer** [comment]
@veyndan i have created a PR #1595, It would be awesome if you can create an issue to put an idea about providing more options for `StateStore` 🙏🏼

**Comment by ahmed3elshaer** [comment]
@JakeWharton @veyndan i think this issue can be closed as #1595 is merged ?
